### PR TITLE
add OfJson and ToJson instances for JsonObject and JsonValue

### PR DESCRIPTION
**Pull request by mt-caret**
Addresses https://github.com/mausch/Fleece/issues/70. 

@gusty I have a background in FP (OCaml, Haskell, and the like), but I'm a novice in F#, so I'm not 100% confident if this works (at least it seems to compile).

### CODE REVIEW

**Comment by gusty** [comment]
Actually, there is already a failing test:

```
From JSON/nested item: Failed: 
item
Expected: Ok (NestedItem {Id = 1;
                Brand = "Sony";
                Availability = Some "1 week";})
Actual: Error
  (JsonTypeMismatch
     (System.Tuple`2[System.String,FSharp.Data.JsonValue][],
      {
  "brand": "Sony",
  "availability": "1 week"
},Array,Object))
 (00:00:00.3021316)
todo: {
  "a": true,
  "b": 200,
  "c": "a"
} ~ {
  "a": true,
  "b": 200,
  "c": null
}
63 tests run: 62 passed, 0 ignored, 1 failed, 0 errored (00:00:17.8344018)
```

and this is coming from the FSharp.Data flavor

I suspect this is due to the fact I mentioned initially, that in FSharp.Data it's an alias:

`type JsonObject = (string * JsonValue) []`

But let's try to make sense of this and evaluate possible workarounds.

**Comment by mt-caret** [comment]
This seems to be an odd error, since the JsonValue OfJson instance seems to be triggering the test failure (all tests pass if that line is commented out).

**Comment by gusty** [comment]
I'll have a look at this and see how can I help to unblock you.

**Comment by gusty** [comment]
It's funny that in Debug mode it works. The problem shows only when compiled in Release mode which means that there is a difference in overload resolution, however I wouldn't be surprised if that's the case.

**Comment by gusty** [comment]
OK, let's add some details of the problem:

As I said, the root problem is the type alias in FSharp.Data.

We can take the failing code and break it down:

```
let (Ok jsonv) = fst jsonValueToTextCodec """{"id": 1, "blah": {"brand": "Sony", "availability": "1 week"}}""" ;;

val jsonv : FSharp.Data.JsonValue =
  {
  "id": 1,
  "blah": {
    "brand": "Sony",
    "availability": "1 week"
  }
}
```
So that's our jsonValue, now if we do `ofJson`

```
> let actual: NestedItem ParseResult = ofJson jsonv ;;

[<Struct>]
val actual : NestedItem ParseResult =
  Error
    (JsonTypeMismatch
       (System.Tuple`2[System.String,FSharp.Data.JsonValue][],
        {
  "brand": "Sony",
  "availability": "1 week"
}, Array, Object))

> Result.mapError string actual ;;
[<Struct>]
val it : Result<NestedItem,string> =
  Error
    "Array expected but got Object while decoding {
  "brand": "Sony",
  "availability": "1 week"
} as System.Tuple`2[System.String,FSharp.Data.JsonValue][]"

```

This comes from this overload https://github.com/mt-caret/Fleece/blob/9f9d85bf395ca19ea099a8868402e5a985642e0c/Fleece/Fleece.fs#L713

instead of this one https://github.com/mt-caret/Fleece/blob/9f9d85bf395ca19ea099a8868402e5a985642e0c/Fleece/Fleece.fs#L763

Why ?

Both of them match as both expect the same result type, one expect `array<'t>` and the other one a `JsonValue` but in FSharpData a `JsonValue` is actually  a `array<string,JsonValue>` so the resolution can go to `array<string,JsonValue>` or the more generic `array<'t>`.

I would expect to go to the more generic `array<'t>` as the other one has a `Default1` marker, but as I said the funny thing is that in Debug more it goes the more specific one, despite the marker.

Possible solutions:

 - Reorder the defaults in order to give priority to the new overload. Complexity: medium to high
 - Use another type for JsonValue instead of an alias. This would likely break existing code. Complexity: easy
 - Same as above but as an erased type, then define all JsonValue overloads is the type. Breaking, complexity Medium.
 - Use conditionals to leave out the JsonValue overloads from FSharpData for now

But I would suggest to leave out JsonValue de/serialization overloads from this PR and work in a separate PR, as doing so this can be merged asap.

**Comment by gusty** [comment]
Still, one thing that it's not entirely clear to me, is why the fact of adding the OfJson for `JsonValue` makes this difference, as `JsonValue` itself is not an alias to a primitive type.

It probably has to do with some dark rules in overload resolution.

Now, what I'm thinking is that probably, even without this PR there might be some conflicting cases, because the main issue seems to be aliasing `JsonObject` with `(string * JsonValue) []` which has a different representation.

So, `(string * JsonValue) []` is represented as a `JArray` but if we consider it a JsonObject it should be `JObject` instead.

For all the above, a better solution would be to change the type alias, to a primitive type that serialize to `JObject`, like `Map<string,JsonValue`.

Maybe I should open a separate issue for that. I'm sure that a side effect of solving that is that this PR will work.

File: Fleece/Fleece.fs [hunk 1]
```diff
@@ -873,6 +881,8 @@ module SystemJson =
         static member inline ToJson (t: 'T, _: Default3) = (^T : (static member ToJSON : ^T -> JsonValue) t)
         static member inline ToJson (t: 'T, _: Default2) = (^T : (static member ToJson : ^T -> JsonValue) t)
 
+        static member ToJson (_: JsonObject, _: Default1) = JsonHelpers.jsonOfJsonObject
+        static member ToJson (_: JsonValue, _: Default1) = Success
```
Comment (gusty) [inline Fleece/Fleece.fs:885]: This incorrect. `ToJson` always succeed, no need to specify `Success`

**Comment by gusty** [comment]
@mt-caret  this is ready to go to master and a release would follow shortly.

I just wonder if you're happy with the changes.

**Pull request merged**